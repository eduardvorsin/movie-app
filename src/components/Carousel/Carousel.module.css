.carousel-container::-webkit-scrollbar {
	display: none;
}

.carousel-container {
	scrollbar-width: none;
}