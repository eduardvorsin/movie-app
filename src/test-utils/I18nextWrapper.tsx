import React, { ReactNode } from 'react';
import i18next from '@/i18n/client';
import { I18nextProvider } from 'react-i18next';

type Props = {
	children: ReactNode,
}

const I18nextWrapper = ({
	children,
}: Props) => {
	return (
		<I18nextProvider i18n={i18next}>
			{children}
		</I18nextProvider>
	);
};

export default I18nextWrapper;