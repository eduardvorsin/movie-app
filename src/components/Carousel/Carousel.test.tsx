import { fireEvent, render, screen } from "@testing-library/react";
import Carousel, { Props, getBreakpointFromValue } from "./Carousel";
import i18next from "@/i18n/client";
import I18nextWrapper from "@/test-utils/I18nextWrapper";

const paginationTypes: NonNullable<Props['paginationType']>[] = ['dots', 'progress', 'fraction'];

jest.useFakeTimers();
describe('Carousel tests', () => {
	beforeEach(() => {
		HTMLElement.prototype.scrollTo = jest.fn().mockImplementation(() => {
			const container = screen.getByTestId<HTMLDivElement>('test-carousel-container');
			container.dispatchEvent(new Event('scroll'));
		});
	});

	it('is rendered correctly', async () => {
		render(
			<Carousel
				testId='test-carousel'
			>
				{new Array(10).fill(null).map((_, index) => (
					<div key={index}>index: {index}</div>
				))}
			</Carousel>
		);

		expect(await screen.findByTestId<HTMLDivElement>('test-carousel')).toBeInTheDocument();
	});

	it('when changing the slide, the mock function should be called', () => {
		const mockFn = jest.fn();
		render(
			<Carousel
				testId='test-carousel'
				onSlideChange={mockFn}
				showArrows
			>
				{new Array(10).fill(null).map((_, index) => (
					<div key={index}>index: {index}</div>
				))}
			</Carousel>
		);

		fireEvent.scroll(screen.getByTestId<HTMLDivElement>('test-carousel-container'), { target: { scrollLeft: 500 } });

		expect(mockFn).toHaveBeenCalledTimes(1);
	});

	it('if mousewheel is true, the mock function should be called when changing the slide through the mouse wheel',
		() => {
			const mockFn = jest.fn();
			render(
				<Carousel
					testId='test-carousel'
					onSlideChange={mockFn}
					mousewheel
					showArrows
				>
					{new Array(10).fill(null).map((_, index) => (
						<div key={index}>index: {index}</div>
					))}
				</Carousel>
			);

			fireEvent.wheel(screen.getByTestId<HTMLDivElement>('test-carousel-container'), { deltaY: 100, });

			expect(mockFn).toHaveBeenCalledTimes(1);
		});

	it('with autoplay set to true, the mock function should be called when changing the slide after 1 second', () => {
		const mockFn = jest.fn();
		render(
			<Carousel
				testId='test-carousel'
				onSlideChange={mockFn}
				autoplay
			>
				{new Array(10).fill(null).map((_, index) => (
					<div key={index}>index: {index}</div>
				))}
			</Carousel>
		);

		jest.advanceTimersByTime(1000);

		expect(mockFn).toHaveBeenCalledTimes(1);
	});

	it('is a basic snapshot', async () => {
		render(
			<Carousel
				testId='test-carousel'
			>
				{new Array(10).fill(null).map((_, index) => (
					<div key={index}>index: {index}</div>
				))}
			</Carousel>
		);

		expect(await screen.findByTestId<HTMLDivElement>('test-carousel')).toBeInTheDocument();
	});

	it('is a snapshot with slidesPerView', async () => {
		render(
			<Carousel
				slidesPerView={3}
				testId='test-carousel'
			>
				{new Array(10).fill(null).map((_, index) => (
					<div key={index}>index: {index}</div>
				))}
			</Carousel>
		);

		expect(await screen.findByTestId<HTMLDivElement>('test-carousel')).toBeInTheDocument();
	});

	it('is a snapshot with slidesPerGroup', async () => {
		render(
			<Carousel
				slidesPerGroup={3}
				testId='test-carousel'
			>
				{new Array(10).fill(null).map((_, index) => (
					<div key={index}>index: {index}</div>
				))}
			</Carousel>
		);

		expect(await screen.findByTestId<HTMLDivElement>('test-carousel')).toBeInTheDocument();
	});

	it('is a snapshot with spaceBetween', async () => {
		render(
			<Carousel
				spaceBetween={50}
				testId='test-carousel'
			>
				{new Array(10).fill(null).map((_, index) => (
					<div key={index}>index: {index}</div>
				))}
			</Carousel>
		);

		expect(await screen.findByTestId<HTMLDivElement>('test-carousel')).toBeInTheDocument();
	});

	it('is a snapshot with showArrows', async () => {
		render(
			<Carousel
				showArrows
				testId='test-carousel'
			>
				{new Array(10).fill(null).map((_, index) => (
					<div key={index}>index: {index}</div>
				))}
			</Carousel>
		);

		expect(await screen.findByTestId<HTMLDivElement>('test-carousel')).toBeInTheDocument();
	});

	it('is a snapshot with showPagination', async () => {
		render(
			<Carousel
				showPagination
				testId='test-carousel'
			>
				{new Array(10).fill(null).map((_, index) => (
					<div key={index}>index: {index}</div>
				))}
			</Carousel>
		);

		expect(await screen.findByTestId<HTMLDivElement>('test-carousel')).toBeInTheDocument();
	});

	it('is a snapshot with showScrollShadow', async () => {
		render(
			<Carousel
				showScrollShadow
				testId='test-carousel'
			>
				{new Array(10).fill(null).map((_, index) => (
					<div key={index}>index: {index}</div>
				))}
			</Carousel>
		);

		expect(await screen.findByTestId<HTMLDivElement>('test-carousel')).toBeInTheDocument();
	});

	it.each(paginationTypes)('is a snapshot with paginationType equal to "%s"', async (paginationType) => {
		render(
			<Carousel
				paginationType={paginationType}
				testId='test-carousel'
			>
				{new Array(10).fill(null).map((_, index) => (
					<div key={index}>index: {index}</div>
				))}
			</Carousel>
		);

		expect(await screen.findByTestId<HTMLDivElement>('test-carousel')).toBeInTheDocument();
	});

	it('is a snapshot with noSwiping', async () => {
		render(
			<Carousel
				noSwiping
				testId='test-carousel'
			>
				{new Array(10).fill(null).map((_, index) => (
					<div key={index}>index: {index}</div>
				))}
			</Carousel>
		);

		expect(await screen.findByTestId<HTMLDivElement>('test-carousel')).toBeInTheDocument();
	});

	it('is a snapshot with breakpoints', async () => {
		render(
			<Carousel
				breakpoints={{
					0: {
						slidesPerView: 1,
						slidesPerGroup: 1,
					},
					1024: {
						slidesPerView: 4,
						slidesPerGroup: 2,
					}
				}}
				testId='test-carousel'
			>
				{new Array(10).fill(null).map((_, index) => (
					<div key={index}>index: {index}</div>
				))}
			</Carousel>
		);

		expect(await screen.findByTestId<HTMLDivElement>('test-carousel')).toBeInTheDocument();
	});
});

describe('Carousel integration tests', () => {
	it('localization into English works correctly', async () => {
		await i18next.changeLanguage('en');
		const translation = new RegExp(i18next.t('carouselArrow.direction', { context: 'left' }));

		render(
			<Carousel
				showArrows
				showPagination
				testId='test-carousel'
			>
				{new Array(10).fill(null).map((_, index) => (
					<div key={index}>index: {index}</div>
				))}
			</Carousel>,
			{ wrapper: I18nextWrapper }
		);

		expect(screen.getByText<HTMLSpanElement>(translation)).toBeInTheDocument();
	});

	it('localization into Russian works correctly', async () => {
		await i18next.changeLanguage('ru');
		const translation = new RegExp(i18next.t('carouselArrow.direction', { context: 'left' }));

		render(
			<Carousel
				showArrows
				showPagination
				testId='test-carousel'
			>
				{new Array(10).fill(null).map((_, index) => (
					<div key={index}>index: {index}</div>
				))}
			</Carousel>,
			{ wrapper: I18nextWrapper }
		);

		expect(screen.getByText<HTMLSpanElement>(translation)).toBeInTheDocument();
	});

	it('is a snapshot with English localization', async () => {
		await i18next.changeLanguage('en');

		render(
			<Carousel
				showArrows
				showPagination
				testId='test-carousel'
			>
				{new Array(10).fill(null).map((_, index) => (
					<div key={index}>index: {index}</div>
				))}
			</Carousel>,
			{ wrapper: I18nextWrapper }
		);

		expect(screen.getByTestId<HTMLDivElement>('test-carousel')).toMatchSnapshot();
	});

	it('is a snapshot with Russian localization', async () => {
		await i18next.changeLanguage('ru');

		render(
			<Carousel
				showArrows
				showPagination
				testId='test-carousel'
			>
				{new Array(10).fill(null).map((_, index) => (
					<div key={index}>index: {index}</div>
				))}
			</Carousel>,
			{ wrapper: I18nextWrapper }
		);

		expect(screen.getByTestId<HTMLDivElement>('test-carousel')).toMatchSnapshot();
	});
});

describe('getBreakpointFromValue tests', () => {
	it('if breakpoints is not an object it should return null', () => {
		expect(getBreakpointFromValue(320, undefined)).toBeNull();
	});

	it('if value is less than all breakpoints it should return null', () => {
		const breakpoints = { 1024: { slidesPerView: 2 } };
		expect(getBreakpointFromValue(320, breakpoints)).toBeNull();
	});

	it('if value is greater than or equal to the corresponding breakpoint, it should return the corresponding object', () => {
		const breakpoints = { 1024: { slidesPerView: 1 } };
		expect(getBreakpointFromValue(1200, breakpoints)).toEqual({ slidesPerView: 1 });
	});
});