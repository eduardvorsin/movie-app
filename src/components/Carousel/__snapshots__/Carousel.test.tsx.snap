// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Carousel integration tests is a snapshot with English localization 1`] = `
<div
  aria-live="polite"
  aria-roledescription="carousel"
  class="overflow-hidden relative  "
  data-testid="test-carousel"
  role="region"
>
  <div
    class="flex overflow-x-auto overflow-y-hidden snap-mandatory snap-x no-scrollbar "
    data-testid="test-carousel-container"
    id=":ri:-container"
  >
    <div
      aria-roledescription="slide"
      class="overflow-hidden flex-shrink-0 flex-grow-0"
      role="group"
      style="margin-right: 0px; flex-basis: calc(100% - 0px); scroll-snap-align: start; scroll-snap-stop: always;"
    >
      <div>
        index: 
        0
      </div>
    </div>
    <div
      aria-roledescription="slide"
      class="overflow-hidden flex-shrink-0 flex-grow-0"
      role="group"
      style="margin-right: 0px; flex-basis: calc(100% - 0px); scroll-snap-align: start; scroll-snap-stop: always;"
    >
      <div>
        index: 
        1
      </div>
    </div>
    <div
      aria-roledescription="slide"
      class="overflow-hidden flex-shrink-0 flex-grow-0"
      role="group"
      style="margin-right: 0px; flex-basis: calc(100% - 0px); scroll-snap-align: start; scroll-snap-stop: always;"
    >
      <div>
        index: 
        2
      </div>
    </div>
    <div
      aria-roledescription="slide"
      class="overflow-hidden flex-shrink-0 flex-grow-0"
      role="group"
      style="margin-right: 0px; flex-basis: calc(100% - 0px); scroll-snap-align: start; scroll-snap-stop: always;"
    >
      <div>
        index: 
        3
      </div>
    </div>
    <div
      aria-roledescription="slide"
      class="overflow-hidden flex-shrink-0 flex-grow-0"
      role="group"
      style="margin-right: 0px; flex-basis: calc(100% - 0px); scroll-snap-align: start; scroll-snap-stop: always;"
    >
      <div>
        index: 
        4
      </div>
    </div>
    <div
      aria-roledescription="slide"
      class="overflow-hidden flex-shrink-0 flex-grow-0"
      role="group"
      style="margin-right: 0px; flex-basis: calc(100% - 0px); scroll-snap-align: start; scroll-snap-stop: always;"
    >
      <div>
        index: 
        5
      </div>
    </div>
    <div
      aria-roledescription="slide"
      class="overflow-hidden flex-shrink-0 flex-grow-0"
      role="group"
      style="margin-right: 0px; flex-basis: calc(100% - 0px); scroll-snap-align: start; scroll-snap-stop: always;"
    >
      <div>
        index: 
        6
      </div>
    </div>
    <div
      aria-roledescription="slide"
      class="overflow-hidden flex-shrink-0 flex-grow-0"
      role="group"
      style="margin-right: 0px; flex-basis: calc(100% - 0px); scroll-snap-align: start; scroll-snap-stop: always;"
    >
      <div>
        index: 
        7
      </div>
    </div>
    <div
      aria-roledescription="slide"
      class="overflow-hidden flex-shrink-0 flex-grow-0"
      role="group"
      style="margin-right: 0px; flex-basis: calc(100% - 0px); scroll-snap-align: start; scroll-snap-stop: always;"
    >
      <div>
        index: 
        8
      </div>
    </div>
    <div
      aria-roledescription="slide"
      class="overflow-hidden flex-shrink-0 flex-grow-0"
      role="group"
      style="margin-right: 0px; flex-basis: calc(100% - 0px); scroll-snap-align: start; scroll-snap-stop: always;"
    >
      <div>
        index: 
        9
      </div>
    </div>
  </div>
  <button
    aria-controls=":ri:-container"
    class="p-2 md:p-1 absolute z-100 -translate-y-1/2 bg-neutral-300/80 enabled:hover:bg-neutral-300/[0.85] enabled:active:bg-neutral-300/90 dark:bg-dark-neutral-300/80 dark:enabled:hover:bg-dark-neutral-300/[0.85] dark:enabled:active:bg-dark-neutral-300/90 text-blue-700 enabled:hover:text-blue-800 enabled:active:text-blue-900 dark:text-blue-300 dark:enabled:hover:text-blue-400 dark:enabled:active:text-blue-500 transition-[color,background-color,opacity] duration-150 disabled:opacity-disabled disabled:cursor-not-allowed left-[0.625rem] top-[calc(50%-36px)]"
    disabled=""
  >
    <svg
      class="w-6 h-6 md:w-10 md:h-10"
      viewBox="0 0 32 32"
    >
      <use
        href="/assets/icons/back-arrow.svg#back-arrow"
      />
    </svg>
    <span
      class="sr-only"
    >
      Previous slide
    </span>
  </button>
  <button
    aria-controls=":ri:-container"
    class="p-2 md:p-1 absolute z-100 -translate-y-1/2 bg-neutral-300/80 enabled:hover:bg-neutral-300/[0.85] enabled:active:bg-neutral-300/90 dark:bg-dark-neutral-300/80 dark:enabled:hover:bg-dark-neutral-300/[0.85] dark:enabled:active:bg-dark-neutral-300/90 text-blue-700 enabled:hover:text-blue-800 enabled:active:text-blue-900 dark:text-blue-300 dark:enabled:hover:text-blue-400 dark:enabled:active:text-blue-500 transition-[color,background-color,opacity] duration-150 disabled:opacity-disabled disabled:cursor-not-allowed right-[0.625rem] top-[calc(50%-36px)]"
  >
    <svg
      class="w-6 h-6 md:w-10 md:h-10"
      viewBox="0 0 32 32"
    >
      <use
        href="/assets/icons/forward-arrow.svg#forward-arrow"
      />
    </svg>
    <span
      class="sr-only"
    >
      Next slide
    </span>
  </button>
  <div
    class="w-full flex justify-center items-center text-center text-200 text-neutral-1000 dark:text-dark-neutral-1000 transition-colors duration-150  "
  >
    <span>
      1
    </span>
    <span
      class="mx-1"
    >
      /
    </span>
    <span>
      10
    </span>
  </div>
</div>
`;

exports[`Carousel integration tests is a snapshot with Russian localization 1`] = `
<div
  aria-live="polite"
  aria-roledescription="carousel"
  class="overflow-hidden relative  "
  data-testid="test-carousel"
  role="region"
>
  <div
    class="flex overflow-x-auto overflow-y-hidden snap-mandatory snap-x no-scrollbar "
    data-testid="test-carousel-container"
    id=":rj:-container"
  >
    <div
      aria-roledescription="slide"
      class="overflow-hidden flex-shrink-0 flex-grow-0"
      role="group"
      style="margin-right: 0px; flex-basis: calc(100% - 0px); scroll-snap-align: start; scroll-snap-stop: always;"
    >
      <div>
        index: 
        0
      </div>
    </div>
    <div
      aria-roledescription="slide"
      class="overflow-hidden flex-shrink-0 flex-grow-0"
      role="group"
      style="margin-right: 0px; flex-basis: calc(100% - 0px); scroll-snap-align: start; scroll-snap-stop: always;"
    >
      <div>
        index: 
        1
      </div>
    </div>
    <div
      aria-roledescription="slide"
      class="overflow-hidden flex-shrink-0 flex-grow-0"
      role="group"
      style="margin-right: 0px; flex-basis: calc(100% - 0px); scroll-snap-align: start; scroll-snap-stop: always;"
    >
      <div>
        index: 
        2
      </div>
    </div>
    <div
      aria-roledescription="slide"
      class="overflow-hidden flex-shrink-0 flex-grow-0"
      role="group"
      style="margin-right: 0px; flex-basis: calc(100% - 0px); scroll-snap-align: start; scroll-snap-stop: always;"
    >
      <div>
        index: 
        3
      </div>
    </div>
    <div
      aria-roledescription="slide"
      class="overflow-hidden flex-shrink-0 flex-grow-0"
      role="group"
      style="margin-right: 0px; flex-basis: calc(100% - 0px); scroll-snap-align: start; scroll-snap-stop: always;"
    >
      <div>
        index: 
        4
      </div>
    </div>
    <div
      aria-roledescription="slide"
      class="overflow-hidden flex-shrink-0 flex-grow-0"
      role="group"
      style="margin-right: 0px; flex-basis: calc(100% - 0px); scroll-snap-align: start; scroll-snap-stop: always;"
    >
      <div>
        index: 
        5
      </div>
    </div>
    <div
      aria-roledescription="slide"
      class="overflow-hidden flex-shrink-0 flex-grow-0"
      role="group"
      style="margin-right: 0px; flex-basis: calc(100% - 0px); scroll-snap-align: start; scroll-snap-stop: always;"
    >
      <div>
        index: 
        6
      </div>
    </div>
    <div
      aria-roledescription="slide"
      class="overflow-hidden flex-shrink-0 flex-grow-0"
      role="group"
      style="margin-right: 0px; flex-basis: calc(100% - 0px); scroll-snap-align: start; scroll-snap-stop: always;"
    >
      <div>
        index: 
        7
      </div>
    </div>
    <div
      aria-roledescription="slide"
      class="overflow-hidden flex-shrink-0 flex-grow-0"
      role="group"
      style="margin-right: 0px; flex-basis: calc(100% - 0px); scroll-snap-align: start; scroll-snap-stop: always;"
    >
      <div>
        index: 
        8
      </div>
    </div>
    <div
      aria-roledescription="slide"
      class="overflow-hidden flex-shrink-0 flex-grow-0"
      role="group"
      style="margin-right: 0px; flex-basis: calc(100% - 0px); scroll-snap-align: start; scroll-snap-stop: always;"
    >
      <div>
        index: 
        9
      </div>
    </div>
  </div>
  <button
    aria-controls=":rj:-container"
    class="p-2 md:p-1 absolute z-100 -translate-y-1/2 bg-neutral-300/80 enabled:hover:bg-neutral-300/[0.85] enabled:active:bg-neutral-300/90 dark:bg-dark-neutral-300/80 dark:enabled:hover:bg-dark-neutral-300/[0.85] dark:enabled:active:bg-dark-neutral-300/90 text-blue-700 enabled:hover:text-blue-800 enabled:active:text-blue-900 dark:text-blue-300 dark:enabled:hover:text-blue-400 dark:enabled:active:text-blue-500 transition-[color,background-color,opacity] duration-150 disabled:opacity-disabled disabled:cursor-not-allowed left-[0.625rem] top-[calc(50%-36px)]"
    disabled=""
  >
    <svg
      class="w-6 h-6 md:w-10 md:h-10"
      viewBox="0 0 32 32"
    >
      <use
        href="/assets/icons/back-arrow.svg#back-arrow"
      />
    </svg>
    <span
      class="sr-only"
    >
      Предыдущий слайд
    </span>
  </button>
  <button
    aria-controls=":rj:-container"
    class="p-2 md:p-1 absolute z-100 -translate-y-1/2 bg-neutral-300/80 enabled:hover:bg-neutral-300/[0.85] enabled:active:bg-neutral-300/90 dark:bg-dark-neutral-300/80 dark:enabled:hover:bg-dark-neutral-300/[0.85] dark:enabled:active:bg-dark-neutral-300/90 text-blue-700 enabled:hover:text-blue-800 enabled:active:text-blue-900 dark:text-blue-300 dark:enabled:hover:text-blue-400 dark:enabled:active:text-blue-500 transition-[color,background-color,opacity] duration-150 disabled:opacity-disabled disabled:cursor-not-allowed right-[0.625rem] top-[calc(50%-36px)]"
  >
    <svg
      class="w-6 h-6 md:w-10 md:h-10"
      viewBox="0 0 32 32"
    >
      <use
        href="/assets/icons/forward-arrow.svg#forward-arrow"
      />
    </svg>
    <span
      class="sr-only"
    >
      Следующий слайд
    </span>
  </button>
  <div
    class="w-full flex justify-center items-center text-center text-200 text-neutral-1000 dark:text-dark-neutral-1000 transition-colors duration-150  "
  >
    <span>
      1
    </span>
    <span
      class="mx-1"
    >
      /
    </span>
    <span>
      10
    </span>
  </div>
</div>
`;
